src/Examples/SequenceToSequence.cs

```diff
@@ -171,43 +167,43 @@ private static double evaluate(TorchTensor eval_data, TransformerModel model, Lo
             return total_loss / eval_data.shape[0];
         }
 
-        static TorchTensor ProcessInput(IEnumerable<string> iter, Func<string, IEnumerable<string>> tokenizer, TorchText.Vocab.Vocab vocab)
+        static Tensor ProcessInput(IEnumerable<string> iter, Func<string, IEnumerable<string>> tokenizer, TorchText.Vocab.Vocab vocab)
         {
-            List<TorchTensor> data = new List<TorchTensor>();
+            List<Tensor> data = new List<Tensor>();
             foreach (var item in iter) {
                 List<long> itemData = new List<long>();
                 foreach (var token in tokenizer(item)) {
                     itemData.Add(vocab[token]);
                 }
                 data.Add(Int64Tensor.from(itemData.ToArray()));
             }
-            return data.Where(t => t.NumberOfElements > 0).ToArray().cat(0);
+            return torch.cat(data.Where(t => t.NumberOfElements > 0).ToArray(), 0);
         }
 
-        static TorchTensor Batchify(TorchTensor data, int batch_size)
+        static Tensor Batchify(Tensor data, int batch_size)
         {
             var nbatch = data.shape[0] / batch_size;
             var d2 = data.narrow(0, 0, nbatch * batch_size).view(batch_size, -1).t();
             return d2.contiguous();
         }
 
-        static (TorchTensor, TorchTensor) GetBatch(TorchTensor source, int index, int bptt)
+        static (Tensor, Tensor) GetBatch(Tensor source, int index, int bptt)
         {
             var len = Math.Min(bptt, source.shape[0] - 1 - index);
-            var data = source[TorchTensorIndex.Slice(index, index + len)];
-            var target = source[TorchTensorIndex.Slice(index + 1, index + 1 + len)].reshape(-1);
+            var data = source[TensorIndex.Slice(index, index + len)];
+            var target = source[TensorIndex.Slice(index + 1, index + 1 + len)].reshape(-1);
             return (data, target);
         }
 
         class TransformerModel : CustomModule
         {
-            private TransformerEncoder transformer_encoder;
+            private impl.TransformerEncoder transformer_encoder;
```

Comment:
What's `impl`?

Reply:
Well... This is one point that needs some discussion.

Generally, in order to mimic Python's new-free constructor syntax for creating torch modules, TorchSharp used to have a special static class 'Modules' in with nothing but factories, all having the same name as the module class:

```csharp
namespace TorchSharp.NN

  public class Conv1d { ... }

  public static partial class Modules { public Conv1d Conv1d(...) { ... } }

}
```

When attempting to move as close to Pytorch as possible, you would want the class and the factory accessible as `torch.conv1d`:

```csharp
namespace TorchSharp

  public static partial class Modules { 

    public class Conv1d { ... }

    public Conv1d Conv1d(...) { ... } 
  }
}
```

but that's not legal. So, there's a choice to be had between placing the factory method in a nested scope, or the type, or forgo the factory altogether and make users say 'new Conv1d()' when creating a module in C#.

Since you rarely __have__ to use the class name, I thought it would make sense to put that in a nested scope rather than the factory. I could certainly be persuaded that relying on C# constructor syntax is just fine, but that does take us away from the "as close to Pytorch as possible" principle.

In most situations, you can infer the type, or just rely on 'Module' as the type name, but in models where the forward method use for a module isn't the standard (and overridden) 'forward(Tensor input)' but something else, you need to capture the most-derived class name in the field declaration. That should be fairly rare.

Reply:
@NiklasGustafsson @dsyme 
Another reason why I am supporting pytorch like naming for TorchSharp is to eventually use a **Python2Csharp** converter to speed up the porting of Pytorch to TorchSharp with a list of Rosyln rules to assist the conversion. 

A successful conversion is one which can be compiled and then using additional Rosyln rules to determine if the results are as expected from the pyTorch through test driven development (TDD). 

What ever final naming or layout of codes we finally reach a consensus, I hope, will make writing the Rosyln rules easier and the TDD simpler.

Reply:
Could you make it `Module.TransformerEncoder` or `Modules.TransformerEncoder` or `ModuleTypes.TransformEncoder`?  At least something that doesn't feel like it's leaking implementation internals (which is what `impl` says to me)

Reply:
I like `Modules.TransformerEncoder`

Reply:
Cool